iox_catalog/src/postgres.rs

```diff
@@ -0,0 +1,874 @@
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ColumnSchema {
+    /// the column id
+    pub id: i32,
+    /// the column type
+    pub data_type: ColumnType,
+}
+
+impl ColumnSchema {
+    fn is_tag(&self) -> bool {
+        self.data_type == ColumnType::Tag
+    }
+
+    fn matches_field_type(&self, field_value: &FieldValue) -> bool {
+        match (field_value, self.data_type) {
+            (FieldValue::I64(_), ColumnType::I64) => true,
+            (FieldValue::U64(_), ColumnType::U64) => true,
+            (FieldValue::F64(_), ColumnType::F64) => true,
+            (FieldValue::String(_), ColumnType::String) => true,
+            (FieldValue::Boolean(_), ColumnType::Bool) => true,
+            _ => false,
+        }
+    }
+}
+
+/// The column data type
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum ColumnType {
+    I64 = 1,
+    U64 = 2,
+    F64 = 3,
+    Bool = 4,
+    String = 5,
+    Time = 6,
+    Tag = 7,
+}
+
+impl ColumnType {
+    fn as_str(&self) -> &str {
+        match self {
+            ColumnType::I64 => "i64",
+            ColumnType::U64 => "u64",
+            ColumnType::F64 => "f64",
+            ColumnType::Bool => "bool",
+            ColumnType::String => "string",
+            ColumnType::Time => "time",
+            ColumnType::Tag => "tag",
+        }
+    }
+}
+
+impl std::fmt::Display for ColumnType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let s = self.as_str();
+
+        write!(f, "{}", s)
+    }
+}
+
+impl TryFrom<i16> for ColumnType {
+    type Error = ();
+
+    fn try_from(value: i16) -> std::prelude::rust_2015::Result<Self, Self::Error> {
+        match value {
+            x if x == ColumnType::I64 as i16 => Ok(ColumnType::I64),
+            x if x == ColumnType::U64 as i16 => Ok(ColumnType::U64),
+            x if x == ColumnType::F64 as i16 => Ok(ColumnType::F64),
+            x if x == ColumnType::Bool as i16 => Ok(ColumnType::Bool),
+            x if x == ColumnType::String as i16 => Ok(ColumnType::String),
+            x if x == ColumnType::Time as i16 => Ok(ColumnType::Time),
+            x if x == ColumnType::Tag as i16 => Ok(ColumnType::Tag),
+            _ => Err(()),
+        }
+    }
+}
+
+/// Given the lines of a write request and an in memory schema, this will validate the write
+/// against the schema, or if new schema is defined, attempt to insert it into the Postgres
+/// catalog. If any new schema is created or found, this function will return a new
+/// `NamespaceSchema` struct which can replace the passed in one in cache.
+///
+/// If another writer attempts to create a column of the same name with a different
+/// type at the same time and beats this caller to it, an error will be returned. If another
+/// writer adds the same schema before this one, then this will load that schema here.
+pub async fn validate_or_insert_schema(
+    lines: Vec<ParsedLine<'_>>,
+    schema: Arc<NamespaceSchema>,
```

Comment:
This fn doesn't need _ownership_ of `schema` - it just needs to borrow it while the fn runs in order to validate it.

If we change this line to borrow the schema, instead of passing in an Arc:

```suggestion
    schema: &NamespaceSchema,
```

We fix two things:
* No pointless `.clone()` of the schema in the caller to pass an `Arc` to this fn
* A caller that has a bare `NamespaceSchema` doesn't have to wrap it in an `Arc` purely to call this fn

The cool bit is, this still [accepts `Arc<NamespaceSchema>`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e25e86789ba920e1902837d6153d1406) too.